### PR TITLE
3776 manage people / users

### PR DESCRIPTION
**Pull request by elliottjf**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,3 +16,6 @@
     rake db:drop db:setup
     rake dev:fake_data
     rails s
+
+### Creating a test user from the rails console
```
Comment (hooverlunch) [inline README.md]: Nice!

File: app/assets/javascripts/backbone/views/person_view.coffee [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+class MS.Views.PersonView extends Backbone.View
+
+  events:
+    'click #person_has_system_access': 'showHideUserFields'
+
+  showHideUserFields: ->
+    if $('#person_has_system_access').is(':checked')
+      $('#user_fields').show()
+    else
+      $('#user_fields').hide()
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/person_view.coffee]: Pretty much all calls to `$` should be `@$` instead. It just ensures they stay within the scope of the View.

File: app/views/admin/people/_form.html.slim [hunk 3]
```diff
@@ -0,0 +1,125 @@
+= simple_form_for(@person,
+    url: @form_action_url,
```
Comment (hooverlunch) [inline app/views/admin/people/_form.html.slim]: I just realized this shouldn't be necessary. I think if you pass `[:admin, @person]` as the first arg, it should work. Then you can strip this out. If I'm right, please strip out from all of the controllers/forms.

File: app/controllers/admin/people_controller.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    @countries = Country.all
+    @organization_choices = organization_choices
+    @division_choices = division_choices
+    @roles_choices = role_choices
+  end
+
+  def role_choices
+    Person::VALID_DIVISION_ROLES.map do |role|
+      [I18n.t("people.roles.#{role}"), role]
+    end
```
Comment (hooverlunch) [inline app/controllers/admin/people_controller.rb]: Simple form should manually translate options for you automatically. See here: http://stackoverflow.com/a/22581612/2066866

@emitche @linesarefuzzy  Please note for future if you didn't already know.

File: app/helpers/application_helper.rb [hunk 5]
```diff
@@ -1,7 +1,9 @@
 module ApplicationHelper
   # adds "http://" if no protocol present
   def urlify(url)
-    URI(url).scheme ? url : "http://" + url
+    # Is there is a better way to handle malformed data, than passing through
+    # with a rescue block?  Note, the original code was just failing.
+    URI(url).scheme ? url : "http://" + url rescue url
```
Comment (hooverlunch) [inline app/helpers/application_helper.rb]: This seems fine to me, but let's please use string interpolation instead of `+`

File: app/models/legacy/member.rb [hunk 6]
```diff
@@ -19,6 +19,12 @@ def migration_data
         country_id: Country.id_from_name(self.country),
         tax_no: national_id,
     }
+    if access_status
+      data[:has_system_access] = true
+      data[:email] = "#{username}@example.com"
+      data[:password] = data[:password_confirmation] = password
```
Comment (hooverlunch) [inline app/models/legacy/member.rb]: sigh

File: app/models/person.rb [hunk 7]
```diff
@@ -41,6 +43,7 @@ class Person < ActiveRecord::Base
   include Notable
   include MediaAttachable
 
+  VALID_DIVISION_ROLES = [:member, :admin]
```
Comment (hooverlunch) [inline app/models/person.rb]: You can use `%i(member admin)` in cases like this. Doesn't make much diff here, but sometimes nice.

File: app/models/person.rb [hunk 8]
```diff
@@ -50,17 +53,101 @@ class Person < ActiveRecord::Base
...
+  # Transient attributes to fascilitate user managment
+  attr_writer :owning_division_role
+  attr_accessor :password, :password_confirmation
+
+  before_save :update_full_name
+  before_save :update_user
+  after_save :handle_roles
+  after_save :clean_up_passwords
+
+  def update_full_name
```
Comment (hooverlunch) [inline app/models/person.rb]: Methods like this that are called as callbacks should typically be private.

File: app/models/person.rb [hunk 9]
```diff
@@ -2,7 +2,6 @@
 #
```
Comment (hooverlunch) [inline app/models/person.rb]: Would be nice to have a comment describing difference between Person and User...

File: app/models/person.rb [hunk 10]
```diff
@@ -50,17 +53,101 @@ class Person < ActiveRecord::Base
   has_many :secondary_agent_loans, class_name: 'Loan', foreign_key: :secondary_agent_id
   has_many :representative_loans,  class_name: 'Loan', foreign_key: :representative_id
 
-  has_one :user, foreign_key: :profile_id
+  has_one :user, foreign_key: :profile_id, autosave: true
```
Comment (hooverlunch) [inline app/models/person.rb]: What happens on destroy? Should this be dependent destroy?

File: app/models/person.rb [hunk 11]
```diff
@@ -50,17 +53,101 @@ class Person < ActiveRecord::Base
...
+    new_role = owning_division_role.present? ? owning_division_role.to_sym : nil
+    # Invalid roles expected to rejected by validation rules. but avoid cryptic error just in case.
+    raise "Unexpected division role: #{new_role}" if new_role && !VALID_DIVISION_ROLES.include?(new_role)
+    if old_role != new_role
+      user.remove_role old_role, division if old_role
+      user.add_role new_role, division if new_role
+    end
+  end
+
+  def clean_up_passwords
```
Comment (hooverlunch) [inline app/models/person.rb]: Just curious, why is this necessary?

File: app/models/person.rb [hunk 12]
```diff
@@ -50,17 +53,101 @@ class Person < ActiveRecord::Base
...
+  def resolve_owning_division_role
+    if user
+      role = user.roles.find_by(resource_type: 'Division', resource_id: division_id)
+      return role.name.to_sym if role
+    end
+    nil
+  end
+
+  def user_required?
+    has_system_access
```
Comment (hooverlunch) [inline app/models/person.rb]: Can/should add `?` at end of this line by convention. Rails supports automatically.

File: app/models/user.rb [hunk 13]
```diff
@@ -76,7 +76,7 @@ def has_some_access?
   # Require a user to have access to at least some division in order to login.
   # Note, this avoids needing to worry about a nil current_division in the controller logic.
   def active_for_authentication?
-    has_some_access?
+    has_some_access? && profile && profile.has_system_access
```
Comment (hooverlunch) [inline app/models/user.rb]: `?` at end of line here also.

File: app/policies/person_policy.rb [hunk 14]
```diff
@@ -1,2 +1,24 @@
...
+  end
+
+  def update_password?
+    user_is_record? || division_admin
+  end
+
+  private
+
+  def user_is_record?
+    @record.user && @user && @user.id != nil && @record.user.id == @user.id
```
Comment (hooverlunch) [inline app/policies/person_policy.rb]: Perhaps `@user == @record.user` would suffice here?

File: app/views/admin/people/_form.html.slim [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+    = f.input_field :website
+
+  = f.input :contact_notes
+    .view-element = @person.contact_notes
+    = f.input_field :contact_notes
+
+  - if policy(@person).update_access?
+
+    = f.input :has_system_access
+      .view-element = @person.has_system_access
```
Comment (hooverlunch) [inline app/views/admin/people/_form.html.slim]: Add `?`

File: app/views/admin/people/_people_grid.slim [hunk 16]
```diff
@@ -0,0 +1,38 @@
+= define_grid(@people_grid, show_filters: :show, hide_csv_button: true,
+    html: {class: 'rowlink table-hover', data: {link: @csv_mode ? nil : 'row'}}) do |g|
+
+    - g.blank_slate t("no_records")
```
Comment (hooverlunch) [inline app/views/admin/people/_people_grid.slim]: Please adjust this to match how I corrected my earlier mistake. See my recent commit to develop.

File: app/views/admin/people/index.html.slim [hunk 17]
```diff
@@ -0,0 +1,51 @@
+- content_for(:title, t("people.shared.people"))
+
+- content_for :with_title
+  = link_to t("people.shared.new"), new_admin_person_path, class: 'btn btn-default new-button'
+
+= render 'layouts/admin/tables'
+
+= render "people_grid"
+
+- unless @people_grid.empty? && !@people_grid.filtering_on?
```
Comment (hooverlunch) [inline app/views/admin/people/index.html.slim]: Same here.

File: app/views/admin/people/new.html.slim [hunk 18]
```diff
@@ -0,0 +1,6 @@
+- content_for :title
+= t("people.shared.new")
```
Comment (hooverlunch) [inline app/views/admin/people/new.html.slim]: This can/should be one line I believe. See other examples in app. (I know it's not consistent but we should move toward one line (@emitche please note)

File: app/views/admin/people/show.html.slim [hunk 19]
```diff
@@ -0,0 +1,40 @@
+- content_for :title
+  = @person.name
```
Comment (hooverlunch) [inline app/views/admin/people/show.html.slim]: Same here.

File: app/views/admin/people/show.html.slim [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+- content_for :styles
+  = stylesheet_link_tag "admin/media/media", media: "all"
```
Comment (hooverlunch) [inline app/views/admin/people/show.html.slim]: This should not be necessary as we are now including these globally. Please try without and if it works, please also remove from organizations#show template.

File: app/views/admin/people/show.html.slim [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+          = t(".delete")
+    .clearfix
+
+  section.block
+    = render 'form'
+
+  section.block
+    h1 = t('people.shared.media')
+    .row 
+      = render "admin/media/index", media: @person.media, owner: @person
```
Comment (hooverlunch) [inline app/views/admin/people/show.html.slim]: Above two lines can be combined into one.

File: app/views/admin/people/show.html.slim [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+
+  section.block
+    h1 = t(:notes)
+
+  javascript:
+    $(function() { new MS.Views.ShowEditView({ el: '.person.details' }) });
```
Comment (hooverlunch) [inline app/views/admin/people/show.html.slim]: If there is a media section, shouldn't there also be a MediaView instance?

File: db/migrate/20160517195315_add_has_system_access_to_people.rb [hunk 23]
```diff
@@ -0,0 +1,5 @@
+class AddHasSystemAccessToPeople < ActiveRecord::Migration
+  def change
+    add_column :people, :has_system_access, :boolean, default: false, null: false
```
Comment (hooverlunch) [inline db/migrate/20160517195315_add_has_system_access_to_people.rb]: Nice constraints.

File: app/policies/person_policy.rb [hunk 24]
```diff
@@ -1,2 +1,24 @@
...
+  end
+
+  class Scope < DivisionOwnedScope
+  end
```
Comment (hooverlunch) [inline app/policies/person_policy.rb]: I'm surprised the person_spec is still passing here since it hasn't been updated. It should probably be updated to cover this new code.

File: app/models/person.rb [hunk 25]
```diff
@@ -50,17 +53,101 @@ class Person < ActiveRecord::Base
...
+  def handle_roles
+    old_role = resolve_owning_division_role
+    new_role = owning_division_role.present? ? owning_division_role.to_sym : nil
+    # Invalid roles expected to rejected by validation rules. but avoid cryptic error just in case.
+    raise "Unexpected division role: #{new_role}" if new_role && !VALID_DIVISION_ROLES.include?(new_role)
+    if old_role != new_role
+      user.remove_role old_role, division if old_role
+      user.add_role new_role, division if new_role
+    end
+  end
```
Comment (hooverlunch) [inline app/models/person.rb]: Seems like it would be good to test all this role resolving and handling code, and perhaps could be done with just a couple of simple tests.

**Comment by hooverlunch** [comment]
Great work here Joseph! I'm super happy with this. A bunch of small notes and I think we need a few model specs in select places. Then I think it's good to go from my end! Roxana will be doing a functional review.

File: app/views/admin/people/_form.html.slim (hunk 3 above)
Comment (elliottjf) [inline app/views/admin/people/_form.html.slim]: Yep.  That seems to work just fine.  Nice!

File: app/models/person.rb (hunk 11 above)
Comment (elliottjf) [inline app/models/person.rb]: I saw similar code in the default devise implementation, so I figured it was a reasonably paranoid thing to do.

Comment (hooverlunch) [inline app/models/person.rb]: Ok that's fine.

On Thu, May 19, 2016 at 11:24 AM, Joseph Elliott <notifications@github.com>
wrote:

> In app/models/person.rb
> <https://github.com/sassafrastech/madeline_system/pull/84#discussion_r63899202>
> :
>
> >
> > +  def handle_roles
> > +    old_role = resolve_owning_division_role
> > +    new_role = owning_division_role.present? ? owning_division_role.to_sym : nil
> > +    # Invalid roles expected to rejected by validation rules. but avoid cryptic error just in case.
> > +    raise "Unexpected division role: #{new_role}" if new_role && !VALID_DIVISION_ROLES.include?(new_role)
> > +    if old_role != new_role
> > +      user.remove_role old_role, division if old_role
> > +      user.add_role new_role, division if new_role
> > +    end
> > +  end
> > +
> > +  def clean_up_passwords
>
> I saw similar code in the default devise implementation, so I figured it
> was a reasonably paranoid thing to do.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/sassafrastech/madeline_system/pull/84/files/4cd7c3742da69348808dcc0253537c94ed299a43#r63899202>
>



-- 
Tom Smyth

Worker-Owner, Sassafras Tech Collective
Specializing in innovative, usable tech for social change
sassafras.coop *·* @sassafrastech

Resident, Touchstone Cohousing
touchstonecohousing.org

File: app/policies/person_policy.rb (hunk 14 above)
Comment (elliottjf) [inline app/policies/person_policy.rb]: I was wanting to guarantee to never match when `@record.user` was nil, as I've been burned in the past in a similar context.  My tendency is to code a bit defensively, but I agree that it's probably not strictly needed here.  I'll simplify.

File: app/views/admin/people/new.html.slim (hunk 18 above)
Comment (elliottjf) [inline app/views/admin/people/new.html.slim]: Okay.  I cleaned up similar usages for orgs and divisions.

File: app/views/admin/people/show.html.slim (hunk 20 above)
Comment (elliottjf) [inline app/views/admin/people/show.html.slim]: Yep. Works fine without. Removed from both here and orgs

File: app/views/admin/people/show.html.slim (hunk 22 above)
Comment (elliottjf) [inline app/views/admin/people/show.html.slim]: It seems to be getting generically included via media/_index.html

File: app/models/person.rb (hunk 25 above)
Comment (elliottjf) [inline app/models/person.rb]: Unit tests added now for both Person and PersonPolicy

**Comment by rgalusca** [comment]
Bug in People, both in the list view and when filtering by city.
example

The same issue appears in Loans and Coops. 
city

**Comment by rgalusca** [comment]
When sorting by city, records are not alphabetical: Brooklyn comes before Amsterdam. The rest of the list seems fine. Sorting by any other category works fine.

**Comment by elliottjf** [comment]
Okay.  The city sorting is a data issue.  There is a leading space.  I'll just add `strip`s when migrating the data.  Regarding the Spanish locale, there should probably be a pass flushing out those translations before we start trying to debug individual display issues.

**Comment by hooverlunch** [comment]
I agree about the Spanish stuff. I wouldn't worry about any of that yet.

On Fri, May 20, 2016 at 11:42 AM, Joseph Elliott <notifications@github.com>
wrote:

> Okay. The city sorting is a data issue. There is a leading space. I'll
> just add strips when migrating the data. Regarding the Spanish locale,
> there should probably be a pass flushing out those translations before we
> start trying to debug individual display issues.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/sassafrastech/madeline_system/pull/84#issuecomment-220641568>
>



-- 
Tom Smyth

Worker-Owner, Sassafras Tech Collective
Specializing in innovative, usable tech for social change
sassafras.coop *·* @sassafrastech

Resident, Touchstone Cohousing
touchstonecohousing.org

**Comment by elliottjf** [comment]
Latest from develop is merged and resolved now.  Is this branch ready to merge down?
